File: src/main/java/com/york/asciiartstudio/asciiArt/AsciiArtBuilder.java
package com.york.asciiartstudio.asciiArt;

import java.util.Optional;

public interface AsciiArtBuilder {

    /**
     * Get the width of the art in characters.
     * @return the width of the art in characters.
     */
    int getWidth();

    /**
     * Get the height of the image in characters.
     * @return the height of the image in characters.
     */
    int getHeight();

    /**
     * @return the max char width possible for the given image source.
     */
    int getMaxCharWidth();

    /**
     * Get how wide each character is in pixels (from the original image).
     * @return Width of each character in pixels
     */
    int getCharWidth();

    /**
     * Resets the builder to the default values.
     */
    AsciiArtBuilder reset();

    /**
     * Set the width of each character in pixels.
     * @param charWidth Width of each character in pixels
     */
    AsciiArtBuilder setCharWidth(int charWidth);

    /**
     * @return the palette of chars currently being used to compose the art.
     */
    String getPalette();

    /**
     * Set the palette of chars to be used to compose the art.
     * @param palette The new palette of chars
     */
    AsciiArtBuilder setPalette(String palette);

    /**
     * @return the art's name if it has one.
     */
    Optional<String> getName();

    /**
     * Set the name of the image.
     * @param name The new name of the image
     */
    AsciiArtBuilder setName(String name);

    /**
     * @return true if shading is inverted, false if not.
     */
    boolean getInvertedShading();

    /**
     * Set whether shading is inverted or not.
     * @param invertedShading true to invert shading, false otherwise
     */
    AsciiArtBuilder setInvertedShading(boolean invertedShading);

    /**
     * @return true if the art is using the default palette, false if not.
     */
    boolean isUsingDefaultPalette();

}

File: src/main/java/com/york/asciiartstudio/model/asciiArt/AsciiArtBuilder.kt
package com.york.asciiartstudio.model.asciiArt

interface AsciiArtBuilder {

    /**
     * Get the width of the art in characters.
     * @return the width of the art in characters.
     */
    val width: Int

    /**
     * Get the height of the image in characters.
     * @return the height of the image in characters.
     */
    val height: Int

    /**
     * @return the max char width possible for the given image source.
     */
    val maxCharWidth: Int

    /**
     * get how wide each character is in pixels (from the original image).
     * @return Width of each character in pixels
     */
    val charWidth: Int

    /**
     * @return the palette of chars currently being used to compose the art.
     */
    val palette: String

    /**
     * @return The name of the image; or null if there is none.
     */
    val name: String?

    /**
     * @return an array of ascii art strings (will be length 1 if building an image).
     */
    val result: Array<String?>

    /**
     * @return true if shading is inverted, false if not.
     */
    val invertedShading: Boolean

    /**
     * @param newCharWidth
     * @return a reference to `this` so that setters can be chained together (ex: builder pattern).
     * @throws IllegalArgumentException if newCharWidth < 1, newCharWidth is greater than the image width, or if newCharWidth * 2 is greater than the image height.
     */
    @Throws(IllegalArgumentException::class)
    fun setCharWidth(newCharWidth: Int): AsciiArtBuilder

    /**
     * @param newPalette A string composed of characters the user wishes to make the art out of. newPalette.length() must be divisible by 256.
     * @return a reference to `this` so that setters can be chained together (ex: builder pattern).
     * @throws IllegalArgumentException if newPalette is not divisible by 256.
     */
    @Throws(IllegalArgumentException::class)
    fun setPalette(newPalette: String): AsciiArtBuilder

    /**
     * @param invertShading
     * @return a reference to `this` so that setters can be chained together (ex: builder pattern).
     */
    fun setInvertedShading(invertShading: Boolean): AsciiArtBuilder

    /**
     * sets the name of the ascii art object.
     * @param newName the new name of the object.
     * @return a reference to `this` so that setters can be chained together (ex: builder pattern).
     */
    fun setName(newName: String?): AsciiArtBuilder

    /**
     * @return true if the art is using the default palette, false if not.
     */
    fun usingDefaultPalette(): Boolean
}